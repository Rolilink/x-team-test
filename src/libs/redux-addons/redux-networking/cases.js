import { createCases } from '../redux-cases';
import networkFactory from './factory';

const { setSuccess, setError, setFetching } = createCases({
  setSuccess: state => ({ ...state, network: networkFactory() }),
  setError: (state, { payload }) => ({ ...state, network: networkFactory({ error: payload }) }),
  setFetching: state => ({ ...state, network: networkFactory({ isFetching: true }) }),
});

export default {
  setSuccess,
  setError,
  setFetching,
};
