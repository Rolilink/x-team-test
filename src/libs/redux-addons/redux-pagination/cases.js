import { createCases } from '../redux-cases';
import paginationFactory from './factory';

const { setPage, setLimit, incrementPage, decrementPage } = createCases({
  setPage: (state, { payload }) => ({
    ...state,
    pagination: paginationFactory({ page: payload.pagination.page, limit: state.pagination.limit }),
  }),
  setLimit: (state, { payload }) => ({
    ...state,
    pagination: paginationFactory({ limit: payload.pagination.limit, page: state.pagination.page }),
  }),
  incrementPage: state => ({
    ...state,
    pagination: paginationFactory({ page: state.pagination.page + 1, limit: state.limit }),
  }),
  decrementPage: state => ({
    ...state,
    pagination: paginationFactory({
      page: state.pagination.page < 1 ? 0 : state.pagination.page - 1,
      limit: state.pagination.limit,
    }),
  }),
});

export default {
  setPage,
  setLimit,
  incrementPage,
  decrementPage,
};
