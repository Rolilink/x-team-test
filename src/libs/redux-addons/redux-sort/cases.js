import _ from 'lodash';
import sortingFactory from './factory';
import { createCases } from '../redux-cases';

const { clearSort, setField, unsetField } = createCases({
  clearSort: state => ({ ...state, sort: sortingFactory() }),
  setField: (state, { payload }) =>
    ({
      ...state,
      sort: sortingFactory({ ...state.sort, ...payload.sort }),
    }),
  unsetField: (state, { payload }) => ({
    ...state,
    sort: sortingFactory(_.omit(state.sort, payload.sort.key)),
  }),
});

export default {
  clearSort,
  setField,
  unsetField,
};
