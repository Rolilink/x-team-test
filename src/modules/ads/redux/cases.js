import { createCases } from '../../../libs/redux-addons/redux-cases';

const { addAd, clearAds } = createCases({
  addAd: (state, { payload }) => ({ ...state, ads: [...state.ads, payload.ad] }),
  clearAds: state => ({ ...state, ads: [] }),
});

export {
  addAd,
  clearAds,
};
