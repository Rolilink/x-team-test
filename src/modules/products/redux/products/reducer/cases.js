import { cases as networkCases } from '../../../../../libs/redux-addons/redux-networking';
import { cases as paginationCases } from '../../../../../libs/redux-addons/redux-pagination';
import { createCase } from '../../../../../libs/redux-addons/redux-cases';

export const addProductsToProductsById = (state, { payload }) => ({
  ...state,
  productsById: payload.products.reduce(
    (productsById, product) => ({
      ...productsById,
      [product.id]: product,
    }),
    state.productsById,
  ),
});

const onProductFetchSuccess = createCase([
  networkCases.setSuccess,
  addProductsToProductsById,
  paginationCases.setPage,
]);

const maybeOnFetchSuccess = (state, action) => {
  const stateSort = Object.keys(state.sort)[0];
  const actionSort = Object.keys(action.payload.sort)[0];

  if (stateSort !== actionSort) {
    return state;
  }

  return onProductFetchSuccess(state, action);
};

export const onFetchSuccess = maybeOnFetchSuccess;
