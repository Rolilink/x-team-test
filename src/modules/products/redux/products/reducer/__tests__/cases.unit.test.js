import { addProductsToProductsById, onFetchSuccess } from '../cases';

const products = [
  { id: 1, size: 17, price: 601, face: '(ง⌐□ل͜□)ง', date: 'Thu Nov 02 2017 02:04:51 GMT-0500 (-05)' },
  { id: 2, size: 13, price: 642, face: '(ღ˘⌣˘ღ)', date: 'Sat Oct 28 2017 05:16:46 GMT-0500 (-05)' },
  { id: 3, size: 13, price: 642, face: '(ღ˘⌣˘ღ)', date: 'Sat Oct 28 2017 05:16:46 GMT-0500 (-05)' },
  { id: 4, size: 13, price: 642, face: '(ღ˘⌣˘ღ)', date: 'Sat Oct 28 2017 05:16:46 GMT-0500 (-05)' },
];

const initialState = {
  productsById: {
    [products[0].id]: products[0],
    [products[1].id]: products[1],
  },
  sort: {
    name: 'asc',
  },
  pagination: {
    page: 1,
    limit: 2,
  },
  network: {
    isFetching: true,
    error: null,
  },
};

describe('Products.Redux.Products.Cases', () => {
  describe('addProductsToProductsById', () => {
    it('should add action.payload.products to productsById', () => {
      const action = { payload: { products: [products[2], products[3]] } };
      const result = addProductsToProductsById(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('onFetchSuccess', () => {
    const action = { payload: { products: [products[2], products[3]], sort: { name: 'asc' }, pagination: { page: 2 } } };
    const result = onFetchSuccess(initialState, action);

    it('should add action.payload.products to productsById', () => {
      expect(result.productsById).toMatchSnapshot();
    });

    it('should set network state to default', () => {
      expect(result.network).toMatchSnapshot();
    });

    it('should set page to action.payload.pagination.page', () => {
      expect(result.pagination.page).toMatchSnapshot();
    });

    it('should not change state if action.payload.sort is different from actual sort', () => {
      const actionWithDiffSort = { payload: { products: [products[2], products[3]], sort: { size: 'asc' }, pagination: { page: 2 } } };
      const resultWithDiffSort = onFetchSuccess(initialState, actionWithDiffSort);

      expect(resultWithDiffSort).toMatchSnapshot();
    });
  });
});
