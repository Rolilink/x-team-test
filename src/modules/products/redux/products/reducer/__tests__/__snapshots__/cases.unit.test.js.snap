// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products.Redux.Products.Cases addProductsToProductsById should add action.payload.products to productsById 1`] = `
Object {
  "network": Object {
    "error": null,
    "isFetching": true,
  },
  "pagination": Object {
    "limit": 2,
    "page": 1,
  },
  "productsById": Object {
    "1": Object {
      "date": "Thu Nov 02 2017 02:04:51 GMT-0500 (-05)",
      "face": "(ง⌐□ل͜□)ง",
      "id": 1,
      "price": 601,
      "size": 17,
    },
    "2": Object {
      "date": "Sat Oct 28 2017 05:16:46 GMT-0500 (-05)",
      "face": "(ღ˘⌣˘ღ)",
      "id": 2,
      "price": 642,
      "size": 13,
    },
    "3": Object {
      "date": "Sat Oct 28 2017 05:16:46 GMT-0500 (-05)",
      "face": "(ღ˘⌣˘ღ)",
      "id": 3,
      "price": 642,
      "size": 13,
    },
    "4": Object {
      "date": "Sat Oct 28 2017 05:16:46 GMT-0500 (-05)",
      "face": "(ღ˘⌣˘ღ)",
      "id": 4,
      "price": 642,
      "size": 13,
    },
  },
  "sort": Object {
    "name": "asc",
  },
}
`;

exports[`Products.Redux.Products.Cases onFetchSuccess should add action.payload.products to productsById 1`] = `
Object {
  "1": Object {
    "date": "Thu Nov 02 2017 02:04:51 GMT-0500 (-05)",
    "face": "(ง⌐□ل͜□)ง",
    "id": 1,
    "price": 601,
    "size": 17,
  },
  "2": Object {
    "date": "Sat Oct 28 2017 05:16:46 GMT-0500 (-05)",
    "face": "(ღ˘⌣˘ღ)",
    "id": 2,
    "price": 642,
    "size": 13,
  },
  "3": Object {
    "date": "Sat Oct 28 2017 05:16:46 GMT-0500 (-05)",
    "face": "(ღ˘⌣˘ღ)",
    "id": 3,
    "price": 642,
    "size": 13,
  },
  "4": Object {
    "date": "Sat Oct 28 2017 05:16:46 GMT-0500 (-05)",
    "face": "(ღ˘⌣˘ღ)",
    "id": 4,
    "price": 642,
    "size": 13,
  },
}
`;

exports[`Products.Redux.Products.Cases onFetchSuccess should not change state if action.payload.sort is different from actual sort 1`] = `
Object {
  "network": Object {
    "error": null,
    "isFetching": true,
  },
  "pagination": Object {
    "limit": 2,
    "page": 1,
  },
  "productsById": Object {
    "1": Object {
      "date": "Thu Nov 02 2017 02:04:51 GMT-0500 (-05)",
      "face": "(ง⌐□ل͜□)ง",
      "id": 1,
      "price": 601,
      "size": 17,
    },
    "2": Object {
      "date": "Sat Oct 28 2017 05:16:46 GMT-0500 (-05)",
      "face": "(ღ˘⌣˘ღ)",
      "id": 2,
      "price": 642,
      "size": 13,
    },
  },
  "sort": Object {
    "name": "asc",
  },
}
`;

exports[`Products.Redux.Products.Cases onFetchSuccess should set network state to default 1`] = `
Object {
  "error": null,
  "isFetching": false,
}
`;

exports[`Products.Redux.Products.Cases onFetchSuccess should set page to action.payload.pagination.page 1`] = `2`;
